File: python/tutorial4_save_data.py
import numpy as np
import datetime

import os
if os.name == 'nt':
    os.add_dll_directory(os.path.join(os.environ["SENSING_DEV_ROOT"], "bin"))

from ionpy import Node, Builder, Buffer, PortMap, Port, Param, Type, TypeCode

if __name__ == "__main__":

    # The following info can be checked with
    # `arv-tool-0.8 -n "<name of device>" control PixelFormat Width Height PayloadSize`
    width = 1920
    height = 1080
    pixelformat = "Mono8"
    payloadsize = 2074880

    num_device = 1

    # where to save gendc
    save_data_directory = 'tutorial_save_gendc_' + datetime.datetime.now().strftime('%Y%m%d%H%M%S')
    os.mkdir(save_data_directory)

    # pipeline setup
    builder = Builder()
    builder.set_target('host')
    builder.with_bb_module('ion-bb')

    # set port
    payloadsize_p = Port('payloadsize', Type(TypeCode.Int, 32, 1), 0)

    # set params
    num_devices = Param('num_devices', str(num_device))
    frame_sync = Param('frame_sync', 'false')
    realtime_diaplay_mode = Param('realtime_diaplay_mode', 'true')
    output_directory = Param('output_directory', save_data_directory)

    # add a node to pipeline
    node = builder.add("image_io_u3v_gendc")\
        .set_param([num_devices, frame_sync, realtime_diaplay_mode, ])

    node_sensor0 = builder.add("image_io_binary_gendc_saver")\
        .set_iport([node.get_port('gendc')[0], node.get_port('device_info')[0], payloadsize_p, ])\
        .set_param([output_directory, 
                    Param('prefix', 'sensor0-') ])

    # create halide buffer for output port
    terminator0 = node_sensor0.get_port('output')
    output0 = Buffer(Type(TypeCode.Int, 32, 1), ())
    terminator0.bind(output0)

    # bind input values to the input port
    payloadsize_p.bind(payloadsize)

    # if num_devices == 2:
    #     payloadsize1_p = Port('payloadsize1', Type(TypeCode.Int, 32, 1), 0)
    #     node_sensor1 = builder.add("image_io_binary_gendc_saver")\
    #         .set_iport([node.get_port('gendc')[1], node.get_port('device_info')[1], payloadsize1_p, ])\
    #         .set_param([output_directory, 
    #                     Param('prefix', 'sensor1-') ])
    #     terminator1 = node_sensor1.get_port('output')
    #     output1 = Buffer(Type(TypeCode.Int, 32, 1), ())
    #     terminator1.bind(output1)
    #     payloadsize1_p.bind(payloadsize)

    num_run = 0
    print("Press any key to stop saving")
    try:
        while True:
            builder.run()
            num_run += 1
    except KeyboardInterrupt:
        pass

    print( num_run, "frames are saved under", save_data_directory)

File: cpp/src/tutorial4_save_data.cpp
/*

g++ src/tutorial4_save_data.cpp -o tutorial4_save_data  \
-I /opt/sensing-dev/include \
-L /opt/sensing-dev/lib \
-L /opt/sensing-dev/lib/x86_64-linux-gnu \
-lHalide -lion-core -ldl -lpthread

*/
#include <filesystem>
#include <iostream>
#include <exception>
#include <ctime>
#include <sstream>

#include <ion/ion.h>

#ifdef _WIN32
    #include <conio.h>
#else
    #include <sys/ioctl.h>
    #include <termios.h>

    bool _kbhit()
    {
        termios term;
        tcgetattr(0, &term);

        termios term_cpy = term;
        term_cpy.c_lflag &= ~ICANON;
        tcsetattr(0, TCSANOW, &term_cpy);

        int byteswaiting;
        ioctl(0, FIONREAD, &byteswaiting);

        tcsetattr(0, TCSANOW, &term);

        return byteswaiting > 0;
    }
#endif

using namespace ion;

int build_and_process_pipeline(int width, int height, int payloadsize, int num_device, std::string saving_diretctory){
    // pipeline setup
    Builder b;
    b.set_target(ion::get_host_target());
    b.with_bb_module("ion-bb");

    // add BB to pipeline
    Node n = b.add("image_io_u3v_gendc")()
      .set_param(
        Param("num_devices", num_device),
        Param("frame_sync", true),
        Param("realtime_diaplay_mode", false)
      );

    if (num_device == 2){
        int32_t payloadsize1 = payloadsize;
        Node n1 = b.add("image_io_binary_gendc_saver")(n["gendc"][1], n["device_info"][1], &payloadsize)
        .set_param(
            Param("prefix", "gendc1-"),
            Param("output_directory", saving_diretctory)
        );
        Halide::Buffer<int> output1 = Halide::Buffer<int>::make_scalar();
        n1["output"].bind(output1);
    }

    n = b.add("image_io_binary_gendc_saver")(n["gendc"][0], n["device_info"][0], &payloadsize)
        .set_param(
            Param("prefix", "gendc0-"),
            Param("output_directory", saving_diretctory)
        );

    // create halide buffer for output port
    Halide::Buffer<int> output = Halide::Buffer<int>::make_scalar();
    n["output"].bind(output);

    int32_t num_run = 0;

    std::cout << "Hit any key to stop saving" << std::endl;

    while(true){
        b.run();
        ++num_run;

        if (_kbhit()) {
            break;
        }
    }
    std::cout << num_run << " frames are saved under " << saving_diretctory << std::endl;

    return 0;

}

int main(int argc, char* argv[]){
    
    try{
        // Create a saving directory
        std::string saving_directory_prefix = "tutorial_save_gendc_";
        std::time_t now;
        std::time(&now);
        std::stringstream ss;

        #ifdef _MSC_VER
        std::tm tm;
        localtime_s(&tm, &now);
        ss << saving_directory_prefix << std::put_time(&tm, "%Y%m%d%H%M%S");
        #else
        std::tm* tm = std::localtime(&now); 
        ss << saving_directory_prefix << std::put_time(tm, "%Y%m%d%H%M%S");
        #endif

        std::filesystem::create_directory(ss.str());

        // The following info can be checked with
        // `arv-tool-0.8 -n "<name of device>" control PixelFormat Width Height PayloadSize`
        int32_t width = 1920;
        int32_t height = 1080;
        int32_t payloadsize = 2074880;

        int32_t num_device = 2;

        int ret = build_and_process_pipeline(width, height, payloadsize, num_device, ss.str());

    } catch(std::exception& e){
        std::cerr << e.what() << std::endl;
        return 1;
    } catch (const ion::Error& e) {
        std::cerr << e.what() << std::endl;
        return 1;
    }
    return 0;
}